src/LimitErrorOutput.hs

```diff
@@ -0,0 +1,128 @@
+{-# LANGUAGE CPP #-}
+{-# LANGUAGE LambdaCase #-}
+{-# LANGUAGE OverloadedStrings #-}
+{-# LANGUAGE DerivingStrategies #-}
+{-# LANGUAGE GeneralizedNewtypeDeriving #-}
+module LimitErrorOutput (
+  Lines(..)
+, Columns(..)
+, limitToTerminalSize
+#ifdef TEST
+, requiredScreenEstate
+, wcwidth
+, parseLocation
+#endif
+) where
+
+import           Prelude hiding (read)
+import           Imports
+import           Data.IORef
+import           Foreign.C
+import qualified Data.ByteString.Char8 as B
+import qualified Data.Text as T
+import qualified Data.Text.Encoding as T
+import           Test.Hspec.Core.Util (stripAnsi)
+
+newtype Lines = Lines { unLines :: Int }
+  deriving newtype (Eq, Show, Ord, Bounded, Enum, Num, Real, Integral)
+
+newtype Columns = Columns { unColumns :: Int }
+  deriving newtype (Eq, Show, Ord, Bounded, Enum, Num, Real, Integral)
+
+data OutputMode = InformationalOutput | ErrorOutput
+
+limitToTerminalSize :: Lines -> Columns -> (ByteString -> IO ()) -> IO (ByteString -> IO ())
+-- NOTES:
+--
+-- 1. Once we see the first error, we switch to line buffering.  This means
+--    that progress reporting will not properly work after the first error. In
+--    practice this is not an issue as Hspec does not use progress reporting
+--    during error output, only during informational output.  However, it is
+--    important to not use line buffering before the first error as to not
+--    impact progress reporting during informational output.
+--
+-- 2. We never output partial error lines. We either output a given line as a
```

Comment:
What's the reason for this? We have a very verbose json matcher that regularly produces a very long line as the first error output, and then I don't see the error.